app/ui/src/main/java/com/fsck/k9/activity/ChooseFolder.java

```diff
@@ -367,29 +364,171 @@ public void run() {
+            return mFilteredFolders.size();
+        }
+
+        @Override
+        public boolean isEnabled(int item) {
+            return true;
+        }
+
+        @Override
+        public boolean areAllItemsEnabled() {
+            return true;
+        }
+
+        public View getView(int position, View convertView, ViewGroup parent) {
+            if (position <= getCount()) {
+                return  getItemView(position, convertView, parent);
+            } else {
+                Timber.e("getView with illegal position=%d called! count is only %d", position, getCount());
+                return null;
+            }
+        }
+
+        public View getItemView(int itemPosition, View convertView, ViewGroup parent) {
+            FolderInfoHolder folder = getItem(itemPosition);
+            View view;
+            if (convertView != null) {
+                view = convertView;
+            } else {
+                view = View.inflate(ChooseFolder.this, R.layout.choose_folder_list_item, null);
+            }
+
+            FolderViewHolder holder = (FolderViewHolder) view.getTag();
+
+            if (holder == null) {
+                holder = new FolderViewHolder();
+                holder.folderName = view.findViewById(R.id.folder_name);
+                holder.folderIcon = view.findViewById(R.id.folder_icon);
+                holder.folderListItemLayout = view.findViewById(R.id.folder_list_item_layout);
+
+                view.setTag(holder);
+            }
+
+            if (folder == null) {
+                return view;
+            }
+
+            holder.folderName.setText(folder.displayName);
+            holder.folderIcon.setImageResource(folderIconProvider.getFolderIcon(folder.folder.getType()));
 
-    static class FolderDisplayData {
-        final long id;
-        final String serverId;
-        final String displayName;
+            if (K9.isWrapFolderNames()) {
+                holder.folderName.setEllipsize(null);
+                holder.folderName.setSingleLine(false);
+            }
+            else {
+                holder.folderName.setEllipsize(TextUtils.TruncateAt.START);
+                holder.folderName.setSingleLine(true);
+            }
 
-        FolderDisplayData(long id, String serverId, String displayName) {
-            this.id = id;
-            this.serverId = serverId;
-            this.displayName = displayName;
+            return view;
         }
 
         @Override
-        public String toString() {
-            return displayName;
+        public boolean hasStableIds() {
+            return true;
+        }
+
+        public Filter getFilter() {
+            return mFilter;
         }
+
+        /**
+         * Filter to search for occurrences of the search-expression in any place of the
+         * folder-name instead of doing just a prefix-search.
+         *
+         * @author Marcus@Wolschon.biz
+         */
+        public class FolderListFilter extends Filter {
```

Comment:
You can probably use `com.fsck.k9.activity.FolderListFilter` instead of copying the code from `FolderList`.